webrender/res/ps_text_run.glsl

```diff
@@ -61,64 +61,82 @@ TextRun fetch_text_run(int address) {
-#endif
+#ifndef WR_FEATURE_GLYPH_TRANSFORM
     // Compute the snapping offset only if the scroll node transform is axis-aligned.
-    if (remove_subpx_offset) {
+    if (transform.is_axis_aligned) {
+#endif
+        // Transform from local space to device space.
+        float device_scale = uDevicePixelRatio / transform.m[3].w;
+        mat2 device_transform = mat2(transform.m) * device_scale;
+
         // Ensure the transformed text offset does not contain a subpixel translation
         // such that glyph snapping is stable for equivalent glyph subpixel positions.
-        vec2 world_text_offset = mat2(transform.m) * text_offset;
-        vec2 device_text_pos = (transform.m[3].xy + world_text_offset) * device_scale;
-        snap_offset += floor(device_text_pos + 0.5) - device_text_pos;
+        vec2 device_text_pos = device_transform * text_offset + transform.m[3].xy * device_scale;
+        snap_offset = floor(device_text_pos + 0.5) - device_text_pos;
+
+        // Snap the glyph offset to a device pixel, using an appropriate bias depending
+        // on whether subpixel positioning is required.
+        vec2 device_glyph_offset = device_transform * glyph_offset;
+        snap_offset += floor(device_glyph_offset + snap_bias) - device_glyph_offset;
+
+        // Transfrom from device space back to local space.
+        mat2 local_transform = inverse(device_transform);
+
+#ifndef WR_FEATURE_GLYPH_TRANSFORM
+        // If not using transformed subpixels, the glyph rect is actually in local space.
+        // So convert the snap offset back to local space.
+        snap_offset = local_transform * snap_offset;
+    }
+#endif
+
+    // Actually translate the glyph rect to a device pixel using the snap offset.
+    glyph_rect.p0 += snap_offset;
 
 #ifdef WR_FEATURE_GLYPH_TRANSFORM
-        // For transformed subpixels, we just need to align the glyph origin to a device pixel.
-        // The transformed text offset has already been snapped, so remove it from the glyph
-        // origin when snapping the glyph.
-        vec2 rough_offset = snap_rect.p0 - world_text_offset * device_scale;
-        snap_offset += floor(rough_offset + snap_bias) - rough_offset;
+    // The glyph rect is in device space, so transform it back to local space.
+    RectWithSize local_rect = transform_rect(glyph_rect, local_transform);
+
+    // Select the corner of the glyph's local space rect that we are processing.
+    vec2 local_pos = local_rect.p0 + local_rect.size * aPosition.xy;
+
+    // If the glyph's local rect would fit inside the local clip rect, then select a corner from
+    // the device space glyph rect to reduce overdraw of clipped pixels in the fragment shader.
+    // Otherwise, fall back to clamping the glyph's local rect to the local clip rect.
+    local_pos = rect_inside_rect(local_rect, local_clip_rect) ?
+                    local_transform * (glyph_rect.p0 + glyph_rect.size * aPosition.xy) :
```

Comment:
would it be straightforward to just clamp that against `local_clip_rect` unconditionally?

Reply:
clamping `local_transform * (glyph_rect.p0 + glyph_rect.size * aPosition.xy)` would result in overdraw? I thought that's the tighter bound alrady

Reply:
I thought you meant to only clamp against the local clip rect. I guess that's possible.